src/nbl/video/IPhysicalDevice.cpp

```diff
@@ -90,4 +90,189 @@ bool IPhysicalDevice::validateLogicalDeviceCreation(const ILogicalDevice::SCreat
+{
+    asset::E_FORMAT smallestTexelBlock = asset::E_FORMAT::EF_UNKNOWN;
+    uint32_t smallestTexelBlockSize = -1;
+
+    auto srcChannels = asset::getFormatChannelCount(srcFormat);
+    if (!srcChannels) return asset::EF_UNKNOWN;
+    auto srcAspects = getImageAspects(srcFormat);
+    auto srcBitDepth = asset::getMaxChannelBitDepth(srcFormat);
+    // TODO "Magic" value of 128.0 until I figure out a value-independent way of comparing precision
+    auto srcPrecision = getFullPrecisionValue(srcFormat, srcChannels);
+
+    // Better way to iterate the bitflags here?
+    for (uint32_t format = 0; format < asset::E_FORMAT::EF_UNKNOWN; format++)
+    {
+        auto f = static_cast<asset::E_FORMAT>(format);
+        if (!validFormats.hasFlags(f))
+        {
+            continue;
+        }
+
+        auto dstChannels = asset::getFormatChannelCount(f);
+        // Verify if promotion is valid from srcFormat -> format
+        bool promotionValid;
+        promotionValid = asset::isBlockCompressionFormat(f) ? f == srcFormat : true; // Can't transcode to compressed formats
+        promotionValid = promotionValid && dstChannels >= srcChannels; // Channel count
+        promotionValid = promotionValid && (getImageAspects(f) & srcAspects).value == srcAspects.value; // Aspects
+        promotionValid = promotionValid && asset::getMaxChannelBitDepth(f) >= srcBitDepth; // Data range
+        promotionValid = promotionValid && getFullPrecisionValue(f, dstChannels) >= srcPrecision; // Precision
+
+        uint32_t texelBlockSize = getTexelOrBlockBytesize(f);
+        if (promotionValid && texelBlockSize < smallestTexelBlockSize)
+        {
+            smallestTexelBlockSize = texelBlockSize;
+            smallestTexelBlock = f;
+        }
+    }
+
+    return smallestTexelBlock;
+}
+
+video::IPhysicalDevice::SFormatBufferUsage convBufferUsage(core::bitflag<asset::IBuffer::E_USAGE_FLAGS> usages)
+{
+    video::IPhysicalDevice::SFormatBufferUsage formatBufUsg;
+    formatBufUsg.isInitialized = 1;
+    formatBufUsg.vertexAttribute = usages.hasFlags(asset::IBuffer::EUF_VERTEX_BUFFER_BIT);
+    formatBufUsg.bufferView = usages.hasFlags(asset::IBuffer::EUF_UNIFORM_TEXEL_BUFFER_BIT);
+    formatBufUsg.storageBufferView = usages.hasFlags(asset::IBuffer::EUF_STORAGE_TEXEL_BUFFER_BIT);
+    // Special flags for these?
+    formatBufUsg.accelerationStructureVertex = usages.hasFlags(asset::IBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT);
+    formatBufUsg.storageBufferViewAtomic = usages.hasFlags(asset::IBuffer::EUF_STORAGE_TEXEL_BUFFER_BIT);
+
+    return formatBufUsg;
+}
+
+asset::E_FORMAT IPhysicalDevice::promoteBufferFormat(const FormatPromotionRequest<asset::IBuffer::E_USAGE_FLAGS> req)
+{
+    auto buf_cache = this->m_formatPromotionCache.buffers;
+    auto cached = buf_cache.find(req);
+    if (cached != buf_cache.end())
+        return cached->second;
+
+    auto srcUsages = convBufferUsage(req.usages);
```

Comment:
do an early out https://github.com/Devsh-Graphics-Programming/Nabla/pull/375/files#r912026997, if you can use the format right away
```cpp
if (req.canUseFormat(getBufferFormatUsages(srcFormat))
{
   buf_cache.insert(cached,{req,srcFormat});
   return srcFormat;
}
```